ocr/utils/sctk-2.4.10/doc/ctmValidator.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.29)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "CTMVALIDATOR 1"
.TH CTMVALIDATOR 1 "2007-06-06" "perl v5.22.1" "Perl Programmers Reference Guide"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ctmValidator.pl \- Script that validates a given CTM file.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&        ctmValidator.pl [\-l <language>] [\-h] \-i <CTM file>
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This Perl program validates a given \s-1CTM\s0 file.
Note that the program will exit after it finds all the syntax errors.
.SH "USAGE"
.IX Header "USAGE"
Examples:
.PP
.Vb 1
\&        ctmValidator.pl \-i ctmfile
.Ve
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fBInput:\fR" 4
.IX Item "Input:"
.RS 4
.PD 0
.IP "\fB\-i <\s-1CTM\s0 file\fR>" 4
.IX Item "-i <CTM file>"
.PD
The \s-1CTM\s0 file that needds to be validated.
.RE
.RS 4
.RE
.IP "\fBOptions:\fR" 4
.IX Item "Options:"
.RS 4
.PD 0
.IP "\fB\-l <language\fR>" 4
.IX Item "-l <language>"
.PD
Check regarding the laguage (default: english).
.IP "\fB\-h\fR" 4
.IX Item "-h"
Display the help.
.RE
.RS 4
.RE
.SH "BUGS"
.IX Header "BUGS"
No known bugs.
.SH "NOTES"
.IX Header "NOTES"
.SH "AUTHORS"
.IX Header "AUTHORS"
.IP "Jerome Ajot <jerome.ajot@nist.gov>" 4
.IX Item "Jerome Ajot <jerome.ajot@nist.gov>"
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Public Domain
.SH "DISCLAIMER"
.IX Header "DISCLAIMER"
This software was developed at the National Institute of Standards and Technology by employees of the Federal Government in the course of their official duties.  Pursuant to Title 17 Section 105 of the United States Code this software is not subject to copyright protection within the United States and is in the public domain. This software is an experimental system. \s-1NIST\s0 assumes no responsibility whatsoever for its use by any party.
.PP
\&\s-1THIS SOFTWARE IS PROVIDED \*(L"AS IS.\*(R"\s0 With regard to this software, \s-1NIST MAKES NO EXPRESS OR IMPLIED WARRANTY AS TO ANY MATTER WHATSOEVER, INCLUDING MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.\s0


ocr/utils/sctk-2.4.10/src/ctmValidator/ctmValidator.pl
#!/usr/bin/perl -w

# CTMVALIDATOR
# Author: Jerome Ajot
#
# This software was developed at the National Institute of Standards and Technology by 
# employees of the Federal Government in the course of their official duties. Pursuant
# to title 17 Section 105 of the United States Code this software is not subject to
# copyright protection and is in the public domain. CTMVALIDATOR is an experimental system.
# NIST assumes no responsibility whatsoever for its use by other parties, and makes no
# guarantees, expressed or implied, about its quality, reliability, or any other
# characteristic. We would appreciate acknowledgement if the software is used.
#
# THIS SOFTWARE IS PROVIDED "AS IS."  With regard to this software, NIST MAKES NO EXPRESS
# OR IMPLIED WARRANTY AS TO ANY MATTER WHATSOEVER, INCLUDING MERCHANTABILITY,
# OR FITNESS FOR A PARTICULAR PURPOSE.

use strict;
use Getopt::Long;
use Data::Dumper;

my $VERSION = "3";

my $USAGE = "\n\n$0 [-l <language>] [-h] -i <CTM file>\n\n".
    "Version: $VERSION\n".
    "Description: This Perl program (version $VERSION) validates a given CTM file.\n".
    "\tNote that the program will exit after it finds all the syntax errors.\n".
    "Options:\n".
    "  -l <language> : check regarding the laguage (default: english)\n".
	"  -h            : print this help message\n".
    "Input:\n".
    "  -i <CTM file>: a CTM file\n\n";

# List the defined types
my %TYPES;
$TYPES{"lex"} = 1;
$TYPES{"frag"} = 1;
$TYPES{"fp"} = 1;
$TYPES{"un-lex"} = 1;
$TYPES{"for-lex"} = 1;
$TYPES{"non-lex"} = 1;
$TYPES{"misc"} = 1;
$TYPES{"noscore"} = 1;

# Number of fields
my $NUM_FIELDS = 8;
my $MIN_FIELDS = 5;

my $inputfile = "";
my $language = "english";

GetOptions
(
    'i=s' => \$inputfile,
    'l=s' => \$language,
    'h'   => sub { print $USAGE; exit },
);

if($inputfile eq "")
{
	print $USAGE;
	exit;
}

open(CTMFILE, $inputfile) or die "Unable to open for read ctm file '$inputfile'";

my $errors = 0;
my $line = 0;

while(<CTMFILE>)
{
	chomp;
	$line++;
	
	# Comments
	# you can have plenty of those
	next if($_ =~ /^;;/);
	
	my @ctm_record = split(/\s+/, $_);
#	print scalar(@ctm_record)." $_\n";
	if(scalar(@ctm_record) < $MIN_FIELDS){
		print "ERROR: [line $line] ctm record must have at least $MIN_FIELDS fields and no whitespace at the beginning\n";
		$errors++;
	} 
	elsif(scalar(@ctm_record) > $NUM_FIELDS)
	{
		print "ERROR: [line $line] ctm record must have no more than $NUM_FIELDS fields and no whitespace at the beginning\n";
		$errors++;
	}
	else
	{

		my $source   = $ctm_record[0];
		my $channel  = $ctm_record[1];
		my $beg_time = $ctm_record[2];
		my $duration = $ctm_record[3];
		my $token    = $ctm_record[4];
		my $conf     = (defined($ctm_record[5]) ? $ctm_record[5] : 0.0);
		my $type     = (defined($ctm_record[6]) ? $ctm_record[6] : "lex");
		my $speaker  = (defined($ctm_record[7]) ? $ctm_record[7] : "spkr");
		
		if($source !~ /^[A-Za-z0-9_-]+$/)
		{
			print "ERROR: [line $line] source '$source' must have alphanumeric, hyphens (-) and underscore (_) characters only\n";
			$errors++;
		}
		
		if($channel !~ /^(\d+|[AB])$/)
		{
			print "ERROR: [line $line] channel '$channel' must be numeric only\n";
			$errors++;
		}
		
		if ($token =~ /(<ALT>|<ALT_BEGIN>|<ALT_END>)/){
		    if($beg_time !~ /^\*$/)
		    {
			print "ERROR: [line $line] begin '$beg_time' time for ALT Tag must be an asterisk /*/\n";
			$errors++;
		    }
		    if($duration !~ /^\*$/)
		    {
			print "ERROR: [line $line] duration '$duration' time for ALT Tag must be an asterisk /*/\n";
			$errors++;
		    }
		} else {
		    if($beg_time !~ /^(\d+(\.\d+)?)$/)
		    {
			print "ERROR: [line $line] begin '$beg_time' time must floating point value\n";
			$errors++;
		    }
	
		    if($duration !~ /^(\d+(\.\d+)?)$/)
		    {
			print "ERROR: [line $line] duration '$duration' time must floating point value\n";
			$errors++;
		    }
		    if( (lc($language) eq "english") && ($type eq "lex") )
		    {
			if( ($token !~ /^[A-Za-z-\']+$/) && ($token !~ /^[A-Za-z]\./) && ($token !~ /@/) && ($token !~ /%(BCN?ACK|HESITATION)/i))
			{
				print "ERROR: [line $line] token '$token' must have alphabetic, hyphens (-) and apostrophes (') characters only\n";
				$errors++;
			}
		    }
		}

		if($conf !~ /^(\d+(\.\d+)?)$/)
		{
			print "ERROR: [line $line] confidence '$conf' time must floating point value\n";
			$errors++;
		}
		
		if(!defined($TYPES{$type}))
		{
			print "ERROR: [line $line] type '$type' is not a known type\n";
			$errors++;
		}
		
		if($speaker !~ /^[A-Za-z0-9_-]+$/)
		{
			print "ERROR: [line $line] speaker '$speaker' must have alphanumeric, hyphens (-) and underscore (_) characters only\n";
			$errors++;
		}
	}
}

close(CTMFILE);

if($errors > 0){
    print "FATAL: Validation '$inputfile' failed due to the $errors previous error(s)\n";
    exit 1;
} else {
    print "Validated $inputfile\n";
    exit 0;
}
